### DISCUSSION TITLE
How to keep track type of a result when I pass clause as param

### DISCUSSION THREAD

**Discussion by tetthys** [prisma/prisma#23516] [Q&A]
### Question

I am developing a custom model class for convenience.

I don't know how to approach from this part.

### Intro

`prismaInstance` is an instance of `PrismaClient`

```
    const result = await prismaInstance.product.findMany({
      select: { id: true },
      take: 10,
    });
```

![스크린샷 2024-03-16 174604](https://github.com/prisma/prisma/assets/115064626/db4b6043-ec09-48a9-8a64-92db1451682e)

As you know, `result` shows correct type.

### Problem

I simplified my class for explanation.

```
import { Prisma, PrismaClient } from "@prisma/client";
import prismaInstance from "../prismaInstance/prismaInstance";

type selectT = Prisma.Args<PrismaClient["product"], "findMany">["select"];

export default class Products {
  private _select?: selectT;
  private _take?: number;

  public select(select: selectT): this {
    this._select = select;
    return this;
  }

  public take(take: number): this {
    this._take = take;
    return this;
  }

  public async get() {
    const result = await prismaInstance.product.findMany({
      select: this._select,
      take: this._take,
    });
    return result;
  }
}

```

![스크린샷 2024-03-16 174626](https://github.com/prisma/prisma/assets/115064626/9d34ddba-c2d8-4309-8a4b-3febd5196d78)

Not surprisingly, it shows a wrong type.

### What I want

I want to follow natural Prisma's type flow for my custom model. That's all.

### Information about Prisma Schema, Client Queries and Environment (optional)

```prisma
// Add your schema.prisma
```

```ts
// Add any relevant Prisma Client queries here
```

- OS: 
- Database: 
- Node.js version: 

- Run `prisma -v` to see your Prisma version and paste it

**Comment by ludralph** [comment]
Hi @tetthys 👋 
Thank you for raising this question. 

You can define the `selectT` type to directly reference the type used by Prisma for the select argument of the findMany method. For example:
```ts

type SelectArgs = Prisma.ProductFindManyArgs['select'];

export default class Products {
  private _select?: SelectArgs;
  private _take?: number;

  public select(select: SelectArgs): this {
    this._select = select;
    return this;
  }

  public take(take: number): this {
    this._take = take;
    return this;
  }

  public async get() {
    const result = await prisma.product.findMany({
      select: this._select,
      take: this._take,
    });
    return result;
  }
}
```

If this answers your question, it would be great if you could **mark this Discussion as answered** to indicate that it has been resolved. 

Otherwise please let us know how else we can help you further or close the Discussion if it was resolved in some other way 🙏

**Reply by tetthys** [reply]
When I type select arguments directly into `prisma.user.findMany()`, VSCode displays the correct filtered type. But I want to implement it in the custom wrapper class. 

**I want to dynamically receive a type of the result, not a type of any clause.**

**Reply by ludralph** [reply]
Hi @tetthys, can you please elaborate with a code example of what you are trying to do?

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that haven’t had any recent activity and appear to be outdated.

If this discussion is still important to you or unresolved, we’d love to hear from you! Feel free to reopen it or start a new one with updated details.

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and being part of the community!
